pkg/controller/elasticquota.go

```diff
@@ -0,0 +1,347 @@
+	eqListerSynced  cache.InformerSynced
+	podListerSynced cache.InformerSynced
+	eqQueue         workqueue.RateLimitingInterface
+	recorder        record.EventRecorder
+}
+
+// NewElasticQuotaController returns a new *ElasticQuotaController
+func NewElasticQuotaController(
+	kubeClient kubernetes.Interface,
+	eqInformer schedinformer.ElasticQuotaInformer,
+	podInformer coreinformer.PodInformer,
+	schedClient schedclientset.Interface,
+	newOpt ...func(ctrl *ElasticQuotaController),
+) *ElasticQuotaController {
+	eventBroadcaster := record.NewBroadcaster()
+	eventBroadcaster.StartRecordingToSink(&corev1.EventSinkImpl{Interface: kubeClient.CoreV1().Events(v1.NamespaceAll)})
+	recorder := eventBroadcaster.NewRecorder(scheme.Scheme, v1.EventSource{Component: "ElasticQuotaController"})
+	// set up elastic quota ctrl
+	ctrl := &ElasticQuotaController{
+		schedClient:     schedClient,
+		eqLister:        eqInformer.Lister(),
+		podLister:       podInformer.Lister(),
+		eqListerSynced:  eqInformer.Informer().HasSynced,
+		podListerSynced: podInformer.Informer().HasSynced,
+		eqQueue:         workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "ElasticQuota"),
+		recorder:        recorder,
+	}
+	for _, f := range newOpt {
+		f(ctrl)
+	}
+	klog.V(5).Info("Setting up elastic quota event handlers")
+	eqInformer.Informer().AddEventHandler(
+		cache.ResourceEventHandlerFuncs{
+			AddFunc:    ctrl.eqAdded,
+			UpdateFunc: ctrl.eqUpdated,
+			DeleteFunc: ctrl.eqDeleted,
+		},
+	)
+	klog.V(5).Info("Setting up pod event handlers")
+	podInformer.Informer().AddEventHandler(
+		cache.ResourceEventHandlerFuncs{
+			AddFunc:    ctrl.podAdded,
+			UpdateFunc: ctrl.podUpdated,
+			DeleteFunc: ctrl.podDeleted,
+		},
+	)
+	return ctrl
+}
+
+func (ctrl *ElasticQuotaController) Run(workers int, stopCh <-chan struct{}) {
+	defer runtime.HandleCrash()
+	defer ctrl.eqQueue.ShutDown()
+	// Start the informer factories to begin populating the informer caches
+	klog.Info("Starting Elastic Quota control loop")
+	// Wait for the caches to be synced before starting workers
+	klog.Info("Waiting for informer caches to sync")
+	if ok := cache.WaitForCacheSync(stopCh, ctrl.eqListerSynced, ctrl.podListerSynced); !ok {
+		klog.Fatalf("Cannot sync caches")
+	}
+	klog.Info("Elastic Quota sync finished")
+	klog.V(5).Infof("Starting %d workers to process elastic quota", workers)
+	// Launch workers to process elastic quota resources
+	for i := 0; i < workers; i++ {
+		go wait.Until(ctrl.worker, time.Second, stopCh)
+	}
+	<-stopCh
+	klog.V(2).Info("Shutting down elastic quota workers")
+}
+
+// WithFakeRecorder will set a fake recorder.It is usually used for unit testing
+func WithFakeRecorder(bufferSize int) func(ctrl *ElasticQuotaController) {
+	return func(ctrl *ElasticQuotaController) {
+		ctrl.recorder = record.NewFakeRecorder(bufferSize)
+	}
+}
+
+func (ctrl *ElasticQuotaController) worker() {
+	for {
+		if quit := ctrl.processNextWorkItem(); quit {
```

Comment:
can ben simplfied to:

```suggestion
	for quit := ctrl.processNextWorkItem(); quit {
```

Reply:
Is this the latest version of go? The 1.15 version of GO does this simplification will report a syntax errors. : )

Reply:
Or like this?
```suggestion
	for  !ctrl.processNextWorkItem() {}
```

Reply:
> Is this the latest version of go?

Ah, I missed a `;` in the end.

> Or like this?

Yes, this is better, but without `!` as we return `true` to quit:

```go
for  ctrl.processNextWorkItem() {}
```

Then it's compatible with the upstream style like https://github.com/kubernetes/kubernetes/blob/ebc87c39d3f453cc3d93c968e0fff2228c37b653/pkg/controller/replicaset/replica_set.go#L518-L521.

Reply:
ok! I've restored it to the original version.